### DISCUSSION TITLE
PSRAM issue

### DISCUSSION THREAD

**Discussion by t-my** [jomjol/AI-on-the-edge-device#856] [General]
Thanks for the project!

I recently bought a esp32-cam from aliexpress (https://www.aliexpress.com/item/1005001642663021.html) and noticed from the logs that there is not enough ram available. 

Is there anything I could do to make the project run on the 4Mb psram esp32? 

Can one run the AI model externally somehow, in linux server for example?

**Comment by jomjol** [comment]
4 MB should work, only 2 MB make a problem. You can use the tflite in any python environment (Keras Tensorflow)

**Reply by t-my** [reply]
Thanks for the quick answer, I was just assuming that we needed more than 4MB of ram since I'm constantly getting this:

```
2022-07-14T12:17:25: =============================================================================================
2022-07-14T12:17:25: =================================== Main Started ============================================
2022-07-14T12:17:25: =============================================================================================
2022-07-14T12:17:25: Not enought PSRAM available. Expected 4.194.304 MByte - available: 2097866
Either not initialzed or too small (2MByte only) or not present at all. Firmware cannot start!!
2022-07-14T12:17:29: task_autodoFlow - next round - Round #1
2022-07-14T12:17:29: 	FlowControll.doFlow - ClassFlowMakeImage
2022-07-14T12:17:29: download_get_handler
2022-07-14T12:17:31: download_get_handler```

**Reply by pos-ei-don** [reply]
So you have the 2MB Version... You need an Upgrade 😉

**Reply by t-my** [reply]
ach so, they sent me wrong unit!

**Reply by friedpa** [reply]
Unfortunately they are cheating most of the time. If you look in the Postings you will see a lot of these comments....

**Comment by t-my** [comment]
Have you considered that the onboarding process (taking the images, cropping the ROI's, enhance contrast and then generating the configuration file) could be done in separate web application using modern tools like react. 

The web app could generate the sd-card contents and perhaps upload it by providing the IP-address of the device. 

Currently the process feels abysmal at best since the device is constantly failing requests and there is no error handling done. The requests just sometimes fail and doing so too often will reboot the device. 

The firmware could be then refactored by "API-first" design and we could build a proper UI layer for the task and that could be served in any normal web server. 

What do you think?

**Reply by friedpa** [reply]
The purpose of the whole Project from jomjol was to get all the necessary program code on a single ESP32. So nothing else is needed, an "external" device would have to be also of the typ of a 24/7 device. Most of the users, I guess do not have a Server.... but this is only my assumption.

**Reply by savikko** [reply]
Also, @t-my,  you should check out the predecessor of this repo: https://github.com/jomjol/water-meter-system-complete

> The interface is a local HTTP-server, that takes an image as input, processes it and gives as an output the readout of the water meter counter, including the subdigits.

**Reply by jomjol** [reply]
I would not recommend to go to "water-meter-system-complete" it is not maintained and is not working with the newest tflite files.

**Reply by jomjol** [reply]
One more comment on the web app etc.:
For a commercial version I can think of a lot of improvements and more professional implementations, including extended error handling, help support etc. 
If someone is interested I'm willing to make an offer, but currently this whole project is covered by a one-man-shown in free time.

**Reply by t-my** [reply]
> Also, @t-my, you should check out the predecessor of this repo: https://github.com/jomjol/water-meter-system-complete
> 
> > The interface is a local HTTP-server, that takes an image as input, processes it and gives as an output the readout of the water meter counter, including the subdigits.



> The purpose of the whole Project from jomjol was to get all the necessary program code on a single ESP32. So nothing else is needed, an "external" device would have to be also of the typ of a 24/7 device. Most of the users, I guess do not have a Server.... but this is only my assumption.

Perhaps I could elaborate a bit more what I visioned.

Web app would mean moving the html files to internet, to Github pages for example. That would not require any hosting or servers from the end user. From the user perspective little would change where one could: 
1. open html page in browser (jomjol.github.com/watermeter etc)
2. provide local ip of the device to the webapp to get the reference image
3. do the configuration 
4. upload the config to the device

This would greatly enhance the user experience. You would not need to wait for such a low end device to serve some static files for you.

**Reply by jomjol** [reply]
Okay - got your your idea. I see two topics with this:
1) the user would need to rely on an "anomyous" web page for giving informations on the water meter, ip-adress etc. --> usually this is not widely accepted
2) I would need to offer a server infrastructure, that needs to be mainained
